glean-core/python/glean/metrics/event.py

```diff
@@ -0,0 +1,199 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+
+import dataclasses
+import json
+from typing import Dict, List, Optional
+
+
+from ..glean import Glean
+from .. import _ffi
+from .._dispatcher import Dispatcher
+from ..testing import ErrorType
+from .. import util
+
+
+from .lifetime import Lifetime
+
+
+@dataclasses.dataclass
+class RecordedEventData:
+    """
+    Deserialized event data.
+    """
+
+    category: str
+    """The event's category, part of the full identifier."""
+
+    name: str
+    """The event's name, part of the full identifier."""
+
+    timestamp: int
+    """The event's timestamp."""
+
+    extra: Optional[Dict[str, str]] = None
+    """Any extra data recorded for the event."""
+
+    @property
+    def identifier(self):
+        if self.category == "":
+            return self.name
+        else:
+            return f"{self.category}.{self.name}"
+
+
+class EventMetricType:
+    """
+    This implements the developer facing API for recording events.
+
+    Instances of this class type are automatically generated by
+    `glean.load_metrics`, allowing developers to record values that were
+    previously registered in the metrics.yaml file.
+
+    The event API only exposes the `EventMetricType.record` method, which
+    takes care of validating the input data and making sure that limits are
+    enforced.
+    """
+
+    def __init__(
+        self,
+        disabled: bool,
+        category: str,
+        lifetime: Lifetime,
+        name: str,
+        send_in_pings: List[str],
+        allowed_extra_keys: List[str],
+    ):
+        self._disabled = disabled
+        self._send_in_pings = send_in_pings
+
+        self._handle = _ffi.lib.glean_new_event_metric(
+            _ffi.ffi_encode_string(category),
+            _ffi.ffi_encode_string(name),
+            _ffi.ffi_encode_vec_string(send_in_pings),
+            len(send_in_pings),
+            lifetime.value,
+            disabled,
+            _ffi.ffi_encode_vec_string(allowed_extra_keys),
+            len(allowed_extra_keys),
+        )
+
+    def __del__(self):
+        if getattr(self, "_handle", 0) != 0:
+            _ffi.lib.glean_destroy_event_metric(self._handle)
+
+    def record(self, extra: Optional[Dict[int, str]] = None):
+        """
+        Record an event by using the information provided by the instance of
+        this class.
+
+        Args:
+            extra (dict of (int, str)): optional. This is a map from keys
+                (which are enumerations) to values. This is used for events
+                where additional richer context is needed. The maximum length
+                for values is 100.
+        """
+        if self._disabled:
+            return
+
+        timestamp = util.time_ms()
+
+        @Dispatcher.launch
+        def record():
```

Comment:
I hope this just shadows the above function definition and thus won't cause any problems?

Reply:
It's a nested function, so not accessible from the outside.  We could really call it whatever we want -- it's never even used by name -- it's just a way to put a function on the dispatcher.  If Python had anonymous functions, I'd use that...

Reply:
👍